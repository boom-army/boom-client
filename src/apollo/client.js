import { ApolloClient, InMemoryCache } from "@apollo/client";
import { offsetLimitPagination } from "@apollo/client/utilities";

export const client = new ApolloClient({
  cache: new InMemoryCache({
    typePolicies: {
      Query: {
        fields: {
          feed: offsetLimitPagination(),
          users: offsetLimitPagination(),
        },
      },
    },
  }),
  uri: process.env.REACT_APP_APOLLO_API || "http://locahost:7777",
  headers: {
    authorization: localStorage.getItem("token") || "",
  },
  context: async ({ req }) => {
    const token = req?.headers?.authorization || null;
    const user= token ? JSON.parse(localStorage.getItem("user")) : null;
    return { user, isLoggedIn: !!token };
  },
});
