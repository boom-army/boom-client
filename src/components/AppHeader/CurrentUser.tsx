import NotificationsActiveIcon from "@mui/icons-material/NotificationsActive";
import NotificationsIcon from "@mui/icons-material/Notifications";
import React from "react";
import { Badge, Box, Grid, IconButton, styled } from "@mui/material";
import { WalletMultiButton } from "../WalletMUIConnect";
import { useNewMentions, useSidebarState } from "../../hooks";

export const CurrentUser: React.FC = () => {
  const { sidebar, toggleRightSidebar } = useSidebarState();
  const { newMentions } = useNewMentions();

  const NotificationPulse = styled(NotificationsActiveIcon)(({ theme }) => ({
    padding: "0.15rem",
    animation: "pulse 5s infinite",
    borderRadius: "50%",
    border: `1px solid ${theme.palette.text.primary}`,
    borderColor: theme.palette.text.primary,
    "@keyframes pulse": {
      "0%": {
        opacity: 1,
      },
      "50%": {
        opacity: 0.5,
      },
      "100%": {
        opacity: 1,
      },
    },
  }));

  return (
    <>
      <Grid
        container
        direction="row"
        justifyContent="flex-start"
        alignItems="center"
      >
        <Box mr={1} display="flex" alignItems="center">
          <WalletMultiButton />
          {sidebar.rightNotificationsFull ? (
            <IconButton
              aria-label="notifications-open"
              onClick={() =>
                toggleRightSidebar(!sidebar.rightNotificationsFull)
              }
            >
              {newMentions?.length ? (
                <Badge badgeContent={newMentions?.length} color="primary">
                  <NotificationPulse />
                </Badge>
              ) : (
                <NotificationPulse />
              )}
            </IconButton>
          ) : (
            <IconButton
              aria-label="notifications"
              onClick={() =>
                toggleRightSidebar(!sidebar.rightNotificationsFull)
              }
            >
              {newMentions?.length ? (
                <Badge badgeContent={newMentions?.length} color="primary">
                  <NotificationsIcon />
                </Badge>
              ) : (
                <NotificationsIcon />
              )}
            </IconButton>
          )}
        </Box>
      </Grid>
    </>
  );
};
