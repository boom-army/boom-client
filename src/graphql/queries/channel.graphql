mutation addChannel(
  $mintAuthority: String!
  $name: String!
  $family: String!
  $description: String
  $image: String
  $status: String
  $channelParentId: String
) {
  addChannel(
    mintAuthority: $mintAuthority
    name: $name
    family: $family
    description: $description
    image: $image
    status: $status
    channelParentId: $channelParentId
  ) {
    id
    mintAuthority
    name
    family
    description
    image
    channelParentId
    status
    membersCount {
      count
      avatars
    }
    createdAt
    updatedAt
  }
}

query channels {
  channels {
    id
    mintAuthority
    name
    family
    description
    image
    channelParentId
    status
    verified
    membersCount {
      count
      avatars
    }
  }
}

mutation channelUnlink($channelId: ID!) {
  channelUnlink(channelId: $channelId)
}

fragment TweetData on Tweet {
  id
  text
  tags
  isTweetMine
  commentsCount
  retweetsCount
  isRetweet
  tipsCount
  createdAt
  channel {
    id
  }
  parentTweet {
    id
    text
    user {
      id
      handle
      avatar
      data {
        avatarIsNFT
      }
    }
  }
  files {
    id
    url
  }
  gif {
    id
    title
    fixedHeightUrl
    originalUrl
  }
  nft {
    id
    publicKey
    name
    symbol
    description
    sellerFeeBasisPoints
    externalUrl
    image
    attributes {
      traitType
      value
    }
    collection {
      name
      family
    }
    properties {
      files {
        uri
        type
      }
      category
      creators {
        address
        share
      }
    }
  }
  user {
    id
    publicAddress
    avatar
    handle
    consumerName
    data {
      avatarIsNFT
    }
  }
  reactions {
    id
    emojiId
    skin
    isMine
    count
  }
}

query channelFeed($channelId: ID!, $offset: Int, $limit: Int) {
  channelFeed(channelId: $channelId, offset: $offset, limit: $limit) {
    ...TweetData
  }
}
