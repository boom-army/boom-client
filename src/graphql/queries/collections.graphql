query getCollection($name: String!) {
  getCollection(name: $name)
}
