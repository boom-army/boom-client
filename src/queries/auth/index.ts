import gql from "graphql-tag";

export const PUBLIC_ADDRESS = gql`
  mutation address($publicAddress: String!) {
    address(publicAddress: $publicAddress) {
      hasPublicAddress
      user {
        id
        nonce
        publicAddress
      }
    }
  }
`;

export const LOGIN_REGISTER = gql`
  mutation loginRegister($publicAddress: String!, $signature: String) {
    loginRegister(publicAddress: $publicAddress, signature: $signature) {
      token
      user {
        id
        handle
        publicAddress
      }
    }
  }
`;
