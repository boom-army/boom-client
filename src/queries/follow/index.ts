import gql from "graphql-tag";

export const FOLLOW = gql`
  mutation follow($id: ID!) {
    follow(id: $id)
  }
`;

export const UNFOLLOW = gql`
  mutation unfollow($id: ID!) {
    unfollow(id: $id)
  }
`;

export const USERS = gql`
  query users($offset: Int, $limit: Int, $filter: String) {
    users(offset: $offset, limit: $limit, filter: $filter) {
      id
      handle
      isFollowing
      isSelf
      consumerName
      avatar
      data {
        avatarMint
        avatarUpdateAuthority
      }
      coverPhoto
      bio
    }
  }
`;

export const USER_FOLLOW = gql`
  query userFollow($offset: Int, $limit: Int) {
    userFollow(offset: $offset, limit: $limit) {
      id
      handle
      avatar
      data {
        avatarMint
        avatarUpdateAuthority
      }
      isFollowing
      isSelf
      consumerName
      bio
    }
  }
`;
