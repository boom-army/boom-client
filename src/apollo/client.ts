import { ApolloClient, InMemoryCache } from "@apollo/client";
import { offsetLimitPagination } from "@apollo/client/utilities";

export const client = new ApolloClient({
  cache: new InMemoryCache({
    typePolicies: {
      Query: {
        fields: {
          feed: offsetLimitPagination(),
          heroFeed: offsetLimitPagination(),
          getChannelById: offsetLimitPagination(),
          users: offsetLimitPagination(),
          searchTweets: offsetLimitPagination(),
          searchUser: offsetLimitPagination(),
          mentions: offsetLimitPagination(),
        },
      },
    },
  }),
  credentials: 'include',
  uri: process.env.REACT_APP_APOLLO_API || "http://locahost:7777",
  headers: {
    authorization: localStorage.getItem("token") || "",
    "Access-Control-Allow-Origin" : "*",
  },
  defaultOptions: {
    watchQuery: {
      context: async ({ req }: any) => {
        const token = req?.headers?.authorization || null;
        const user = token ? localStorage.getItem("user")?.toString() : null;
        return { user, isLoggedIn: !!token };
      },
    },
  },
});
