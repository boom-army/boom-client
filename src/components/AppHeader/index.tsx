import { useCallback, useEffect, useMemo, useContext } from "react";
import base58 from "bs58";
import {
  AppBar,
  Container,
  Grid,
  Toolbar,
} from "@mui/material";
import { CurrentUser } from "../CurrentUser";
import { PUBLIC_ADDRESS, LOGIN_REGISTER } from "../../queries/auth";
import { USER_FOLLOW } from "../../queries/follow";
import { Theme } from "../../contexts/theme";
import { UserContext } from "../../contexts/user";
import { useMutation } from "@apollo/client";
import { useSnackbar } from "../../contexts/snackbar";
import { useWallet } from "@solana/wallet-adapter-react";
import { FeedDocument } from "../../generated/graphql";
import { Box } from "@mui/system";
import { Link } from "react-router-dom";
import { RoutePath } from "../../constants";

export const AppHeader = () => {
  const { connected, wallet, publicKey, signMessage } = useWallet();
  const { enqueueSnackbar } = useSnackbar();
  const { setUser } = useContext(UserContext);

  const token = localStorage.getItem("token");
  const headerImg =
    localStorage.getItem(Theme.StorageTag) === Theme.Light
      ? "/assets/boom-logo-dark.png"
      : "/assets/boom-logo-light.png";

  const [getNonce] = useMutation(PUBLIC_ADDRESS);
  const [setLogin] = useMutation(LOGIN_REGISTER, {
    refetchQueries: [{ query: FeedDocument }, { query: USER_FOLLOW }],
    onCompleted({ loginRegister }) {
      if (localStorage) {
        setUser(loginRegister.user);
        localStorage.setItem("token", loginRegister.token);
      }
    },
  });

  const walletPublicKey = useMemo(() => publicKey?.toBase58(), [publicKey]);

  const signin = useCallback(async () => {
    if (!walletPublicKey) {
      enqueueSnackbar("Wallet not connected!", { variant: "error" });
      return;
    }

    if (!signMessage) {
      enqueueSnackbar("Wallet does not support message signing!", {
        variant: "error",
      });
      return;
    }

    try {
      const {
        data: { address },
      } = await getNonce({
        variables: { publicAddress: walletPublicKey },
      });
      if (address.hasPublicAddress) {
        const data = new TextEncoder().encode(address.user.nonce);
        const signature = await signMessage(data);

        await setLogin({
          variables: {
            publicAddress: walletPublicKey,
            signature: base58.encode(signature),
          },
        });
        window.location.reload();
        enqueueSnackbar(
          `Wallet ${walletPublicKey} connected to account. Happy posting.`,
          { variant: "success" }
        );
      } else {
        await setLogin({ variables: { publicAddress: walletPublicKey } });
        enqueueSnackbar(`Wallet ${walletPublicKey} created for account.`, {
          variant: "success",
        });
      }
    } catch (error) {
      enqueueSnackbar(`Error connecting: ${error}`, { variant: "error" });
    }
  }, [signMessage, getNonce, setLogin, walletPublicKey, enqueueSnackbar]);

  useEffect(() => {
    if (wallet && !token && connected) signin();
  }, [wallet, signin, token, connected]);

  // const logout = () => {
  //   localStorage.clear();
  //   window.location.replace('/');
  // };

  return (
    <AppBar
      position="static"
      sx={{
        borderBottom: (t) => `1px solid ${t.palette.divider}`,
      }}
    >
      <Toolbar variant="dense">
        {/* Remove this when we go full width */}
        <Container maxWidth="lg">
          <Grid container>
            <Grid item xs={12}>
              <Box display="flex" justifyContent="space-between">
                <Box mr={1} display="flex" alignItems="center">
                  <Link
                    to={RoutePath.HOME}
                    style={{ display: "flex", alignItems: "center" }}
                  >
                    <img src={headerImg} alt="Boom" height={20} />
                  </Link>
                </Box>
                <Box mt={0.5}>
                  <CurrentUser />
                </Box>
              </Box>
            </Grid>
          </Grid>
        </Container>
      </Toolbar>
    </AppBar>
  );
};
