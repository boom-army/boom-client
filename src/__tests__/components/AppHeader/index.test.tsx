import React from "react";
import { AppHeader } from "../../../components/AppHeader/index";
import { act, fireEvent, render,screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { SnackbarProvider } from "notistack";

import { MockedProvider } from "@apollo/client/testing";
import { PUBLIC_ADDRESS, LOGIN_REGISTER } from "../../../queries/auth";
import { USER_FOLLOW } from "../../../queries/follow";

let user: any = {
  id: "ckzny1iv10031lmn568gj3b3n",
  publicAddress: "JBZ52cKhHiFJdzQBNWnp3Xy2jHgDgkhWwSLWoNgLNxD4",
  avatar: "",
  handle: "shy-cloud-4965",
  consumerName: "shy-cloud-4965",
  __typename: "User",
};

export const __mocks__: any = [
  {
    request: {
      query: PUBLIC_ADDRESS,
      variables: {},
    },
    result: {
      data: {},
      errors: "An error occurred",
    },
  },
];

describe("<AppHeader/> component :", () => {
  test("display <AppHeader/> component ", async () => {

    //snapshot testing 
    let rendered;
    await act(async () => {
      rendered = render(
        <MockedProvider mocks={__mocks__} addTypename={false}>
          <SnackbarProvider>
            <AppHeader />
          </SnackbarProvider>
        </MockedProvider>,
        { wrapper: MemoryRouter }
      );
    });
    expect(rendered).toMatchSnapshot();
   
});

//assertion testing

test("Assertion testing of <AppHeader/> component ", async () => {
  render(
    <MockedProvider mocks={__mocks__} addTypename={false}>
      <SnackbarProvider>
        <AppHeader />
      </SnackbarProvider>
    </MockedProvider>,
    { wrapper: MemoryRouter }
  );
  expect(screen.getByText('Boom β')).toBeInTheDocument();

  expect(screen.getByText('Select Wallet')).toBeInTheDocument();

  expect(screen.getByRole('link', { name: 'Boom β' })).toHaveAttribute('href', '/')
});
});


