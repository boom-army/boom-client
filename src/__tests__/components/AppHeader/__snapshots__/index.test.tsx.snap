// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AppHeader/> component : display <AppHeader/> component  1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <header
        class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionAbsolute css-1d8ws7u-MuiPaper-root-MuiAppBar-root"
      >
        <div
          class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
        >
          <div
            class="MuiToolbar-root MuiToolbar-dense css-lz9mh9-MuiToolbar-root"
          >
            <div
              class="MuiGrid-root MuiGrid-container css-1c1ljyi-MuiGrid-root"
            >
              <div
                class="MuiBox-root css-12z0wuy"
              >
                <a
                  href="/"
                >
                  <h6
                    class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
                  >
                    Boom β
                  </h6>
                </a>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-container css-ded95z-MuiGrid-root"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiButton-label"
                >
                  Select Wallet
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </div>
      </header>
    </div>
  </body>,
  "container": <div>
    <header
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionAbsolute css-1d8ws7u-MuiPaper-root-MuiAppBar-root"
    >
      <div
        class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
      >
        <div
          class="MuiToolbar-root MuiToolbar-dense css-lz9mh9-MuiToolbar-root"
        >
          <div
            class="MuiGrid-root MuiGrid-container css-1c1ljyi-MuiGrid-root"
          >
            <div
              class="MuiBox-root css-12z0wuy"
            >
              <a
                href="/"
              >
                <h6
                  class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
                >
                  Boom β
                </h6>
              </a>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-container css-ded95z-MuiGrid-root"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiButton-label"
              >
                Select Wallet
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </header>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
